src/bytes.rs

```diff
@@ -994,12 +1043,41 @@ unsafe fn promotable_to_vec(
     }
 }
 
+unsafe fn promotable_to_mut(
+    data: &AtomicPtr<()>,
+    ptr: *const u8,
+    len: usize,
+    f: fn(*mut ()) -> *mut u8,
+) -> BytesMut {
+    let shared = data.load(Ordering::Acquire);
+    let kind = shared as usize & KIND_MASK;
+
+    if kind == KIND_ARC {
+        shared_to_mut_impl(shared.cast(), ptr, len)
+    } else {
+        debug_assert_eq!(kind, KIND_VEC);
+
+        let buf = f(shared);
+        let off = offset_from(ptr, buf);
+        let cap = off + len;
+        let v = Vec::from_raw_parts(buf, cap, cap);
```

Comment:
I don't think this is right. The resulting `BytesMut` will reference the slice `off..v.len()` in the vector, but it's possible that the underlying allocation has more data after the end of this `Bytes`. But since you pass `cap` twice to `Vec::from_raw_parts`, this cannot possibly return a `BytesMut` with more data after the end of the `BytesMut`'s view.

Can you please add a test that would have caught this?

Reply:
I can add a test for sure! I did base myself on the code from `to_vec` where the cap is set as `let cap = (ptr as usize - buf as usize) + len;` which is essentially the same as I do here. If this code leaks then I am pretty sure `to_vec` also leaks.

Reply:
Well, that would be unfortunate. Would you be able to verify whether there's a bug in `to_vec` or not, and if so, submit a bug (or PR) for it?

Reply:
I've concluded that it's not possible to have a KIND_VEC whose view does not end at the end of the allocation. All apis that can cut off the end will promote to KIND_ARC. See e.g. truncate:

https://github.com/tokio-rs/bytes/blob/4e2c9c065a06bf9cb5d7dd46e3b29f62a1c20057/src/bytes.rs#L479-L492

And `impl From<Vec<u8>> for Bytes` also goes straight to KIND_ARC unless `len == cap`.

Reply:
Still, even if it's correct like this, please add a comment mentioning it.